### DISCUSSION TITLE
Wrist rest generation?

### DISCUSSION THREAD

**Discussion by TonyWu20** [joshreve/dactyl-keyboard#54] [Ideas]
Personally, I prefer a wrist rest that follows the shape and curve of the near-user-side, e.g., the following one from diykeyboard:
![wrist_rest](https://www.diykeyboards.com/images/com_hikashop/upload/dactyl_manuform_wrist_rests.jpg)
I am trying to make myself one in blender. I thought it should be some boolean operations to create the shape that can match the interface, but so far haven't made it. So I wonder if it will be applicable to generate them along with the case with some coding about the boolean operations to match the different shapes from thumb cluster variations.

**Comment by joshreve** [comment]
I think this is definitely possible, but a little tricky.  If you just create a geometry you want sticking out for the rest and attempt a boolean, it will likely create 2 solids which can make a mess.  You can try that and see what the engine does.  If it fails, you may have to use the front and thumb walls to generate the front edge of the rest which will definitely get more complicated.  

While I am not a wrist rest user, I know a lot of people are, so I'd be willing to help make this happen if you can make some progress and test out a few approaches first.

**Reply by TonyWu20** [reply]
I’m fiddling with the clojure code from carbonfet’s fork, because I generated cases from it last time. It took me quite a while to get a solidified part “inside” the case, which I plan to use it to do difference operation with the attached new cube. The obstacle is OpenSCAD doesn’t have a method to fill a outlined area. If OpenCASCADE has better support for this I will try it.

**Reply by joshreve** [reply]
If you take a look at the cad query base plate I extract a planar cut off the bottom, inspect it, and build a solid from it.  I didn't originally think of it, but that approach could work in this case.  It wouldn't necessarily hug the geometry higher up, but would match up at the bottom.  You could theoretically build it onto the base plate or something similar.

**Reply by TonyWu20** [reply]
Well, the major difficulty for using Boolean should be the over complex geometries of the generated model. Too many vertices on the model. The Boolean operators failed to produce the desired ideal cut. So I have to find another way: *manually* **extrude the interface outward**. It’s like peeling off the “layer” of the interface, extrude them by duplicating and moving the set of vertices, and then the other side is free to decorate. 
With hours of struggling in clojure and blender, I finally reached a should-be-ok-starting point: 
1. I commented out lines in case-walls sections to leave the necessary walls for my wrist rest interface. It leaves the necessary thumb cluster walls and the mainframe walls extended to the pinky corner only.
2. I exported this piece to Blender. I chose one vertex point at the pinky with some height from the ground, assuming its Z coordinate is the height of my wrist rest solid. Then I duplicated a series of points from the upper end of each vertical edges of the wall surfaces, then set them to the same height as the first one.
3. I connect every four vertices to create a duplicated face. After that, a series of faces are created to “peel them off” from this wall.
4. The next step will be duplicating this set of vertices for these faces, then extrude them in the XY plane. In this way, the created solid will surely be able to attach to the original wall of the keyboard at the interface. The other part can be furnished by other shapes and operations.
5.  Finally: the key step is obviously **step 2**. I think this is code-able. It should be like adjusting the Z offset of the numerous “post” in the hulling sections that responsible for creating the thumb wall and case wall. I will begin to sort out this after I complete my file in Blender.

**Reply by TonyWu20** [reply]
Here’s what I got. To complete the **step 2** there’s a much easier way: just put a cube on top of the wall and do the difference. Then it cuts it all.

<img width="2032" alt="CD7B2F4C-A2E6-4BB2-8DF6-D830E52F5D99" src="https://user-images.githubusercontent.com/45414421/140011422-0d4b1cdc-ed15-4694-8b44-bb20fd856dd5.png">

**Comment by TonyWu20** [comment]
Summarizing what I learned today:
1. The overall height of the wrist can be easily controlled by putting a large cube on top and then difference the necessary back walls with the cube.
2. Since the back wall generation is a union of parts, I think I can get the projection in YZ plane of each part, extrude the projection along Y axis, use hull to connect the extruded cube and the part.

**Comment by TonyWu20** [comment]
Lol, I just realized I made things too complicated. I can check the coordinates of the vertices in the exported STL file (where the language of openscad sucks). So I did this:
![4A00457F-6407-4F78-90FA-736FC272F50B](https://user-images.githubusercontent.com/45414421/140378117-11b84a16-dc12-4338-a785-259fc29a5c12.jpeg)
![3E6C8D2C-5772-46A9-BF3E-4F74F5B231C1](https://user-images.githubusercontent.com/45414421/140378774-60c97119-af61-4948-a95c-c6d844335aa5.jpeg)
Haven’t finished yet, I hope this can be decently chopped off with the original wall in openscad, so I will get the wrist rest with correct interface geometries. I wonder if the access of the target shape properties would be easier in your python code

**Reply by joshreve** [reply]
The target shape is a bit challenging because of the generation structure.  Now that you mentioned projection, you could literally project the whole design to the floor and extrude up to get a flat edge. You can use that to make a large subtract from a wrist rest shape.

**Reply by joshreve** [reply]
Once you project it, if that works, the subtraction should be trivial compared to the hollowed design.

**Reply by joshreve** [reply]
If you want to try the projection, look at the pysolid baseplate.  It would still need to be extruded, but the functions there should give you the outline.

**Reply by TonyWu20** [reply]
Boolean is still not working properly at the interface, probably due to a bunch of messy vertices generated from the series of hull operations.

**Comment by TonyWu20** [comment]
I modify the `wall-brace` and `key-wall-brace` functions in the clojure codes to create one seemingly works shape:  
![image](https://user-images.githubusercontent.com/45414421/140495487-138df069-13d8-42da-aaf6-b9ef7c0eaec4.png)
The modified `wall-brace`:
```
(defn wall-brace-out [place1 dx1 dy1 post1 place2 dx2 dy2 post2]
  (union
   (hull ;connect the original generated part and the translated part
      (hull
       (place1 post1)
       (place1 (translate (wall-locate3 dx1 dy1) post1)); only wall-locate3 to generate the outmost surface of the wall
       (place2 post2)
       (place2 (translate (wall-locate3 dx2 dy2) post2))
       )
     (translate [0 -150 0] ;translate this part by 150 mm along y-axis
      (hull
       (place1 post1)
       (place1 (translate (wall-locate3 dx1 dy1) post1))
       (place2 post2)
       (place2 (translate (wall-locate3 dx2 dy2) post2))
       )
     )
   )
   (hull ;same idea as the previous part
     (bottom-hull
      (place1 (translate (wall-locate3 dx1 dy1) post1))
      (place2 (translate (wall-locate3 dx2 dy2) post2))
     )
     (translate [0 -150 0]
     (bottom-hull
      (place1 (translate (wall-locate3 dx1 dy1) post1))
      (place2 (translate (wall-locate3 dx2 dy2) post2))
     )
     )
   )
  )
)
```
The modified `key-wall-brace`
```
(defn key-wall-brace-out [x1 y1 dx1 dy1 post1 x2 y2 dx2 dy2 post2]
  (wall-brace-out (partial key-place x1 y1) dx1 dy1 post1
              (partial key-place x2 y2) dx2 dy2 post2))
```
Though I'm more familiar with Python, I haven't tried to fully understand your ported code here, so I can't provide the corresponding code snippets to you for testing.

**Reply by TonyWu20** [reply]
https://user-images.githubusercontent.com/45414421/140546727-41481a9b-3d1c-4ac8-b455-69be2671447d.mp4

Current progress. Used some cubes to chop off some messy corners and shapes. 99% percent of the operations are still done by coding. Except this big slanted cube to trim the slope (angle is determined from the vertices coordinates of the parallel edge):
![image](https://user-images.githubusercontent.com/45414421/140547526-ed3c23a6-ca11-45ed-a833-04a90be7d681.png)
